File: IrrgartenJava/IrrgartenJava/src/irrgarten/Player.java
package irrgarten;
import java.util.ArrayList;

/**
 * La clase Player representa a un jugador en el juego. Cada jugador tiene un nombre,
 * número, inteligencia, fuerza, salud, posición en el laberinto, y puede poseer armas y escudos.
 * Puede realizar acciones como moverse, atacar, defenderse, recibir recompensas y más.
 * Además, proporciona una representación en cadena de sus atributos.
 *
 * @author Juan Carlos Vílchez Sánchez
 */
public class Player extends LabyrinthCharacter{

    // Constantes de la clase
    private static final int MAX_WEAPONS = 2;
    private static final int MAX_SHIELDS = 3;
    private static final int INITIAL_HEALTH = 10;
    private static final int HITS2LOSE = 3;

    // Variables de instancia
    private char number;
    private int consecutiveHits = 0;
    private ArrayList<Weapon> weapons;
    private ArrayList<Shield> shields;
    private ShieldCardDeck shieldCardDeck;
    private WeaponCardDeck weaponCardDeck;

    /**
     * Construye un nuevo objeto Player con el número, inteligencia y fuerza especificados.
     *
     * @param number Número del jugador.
     * @param intelligence Nivel de inteligencia del jugador.
     * @param strength Nivel de fuerza del jugador.
     */
    public Player(char number, float intelligence, float strength){
        super("Player# " + number, intelligence, strength, INITIAL_HEALTH);
        this.number = number;
        shields = new ArrayList<>();
        weapons = new ArrayList<>();
        weaponCardDeck = new WeaponCardDeck();
        weaponCardDeck.addCards();
        shieldCardDeck = new ShieldCardDeck();
        shieldCardDeck.addCards();
    }
    
    
    public Player(Player other){
        super(other);
        number = other.number;
        shields = other.shields;
        weapons = other.weapons;
        shieldCardDeck = other.shieldCardDeck;
        weaponCardDeck = other.weaponCardDeck;
    }

    /**
     * Resucita al jugador, restableciendo su salud, escudos y armas.
     */
    public void resurrect(){
        setHealth(INITIAL_HEALTH);
        shields.clear();
        weapons.clear();
        consecutiveHits = 0;
    }

    /**
     * Obtiene el número del jugador.
     *
     * @return El número del jugador.
     */
    public char getNumber(){
        return number;
    }

    /**
     * Realiza un movimiento en la dirección especificada, teniendo en cuenta las direcciones válidas.
     *
     * @param direction Dirección en la que se moverá el jugador.
     * @param validMoves Array de direcciones válidas.
     * @return La dirección final en la que se movió el jugador.
     */
    public Directions move(Directions direction, ArrayList<Directions> validMoves){
        int size = validMoves.size();
        boolean contained = validMoves.contains(direction);
        
        if((size > 0) && (!contained)){
            Directions firstElement = validMoves.get(0);
            return firstElement;
        }else{
            return direction;
        }
    }

    /**
     * Realiza un ataque y devuelve la intensidad del ataque.
     *
     * @return La intensidad del ataque basada en la fuerza del jugador y las armas que posee.
     */
    @Override
    public float attack(){
        return sumWeapons() + getStrength();
    }

    /**
     * Intenta defenderse contra un ataque recibido.
     *
     * @param receivedAttack La intensidad del ataque recibido.
     * @return true si el jugador se defiende con éxito, false en caso contrario.
     */
    @Override
    public boolean defend(float receivedAttack){
        return manageHit(receivedAttack);
    }

    /**
     * Recibe una recompensa después de ganar un combate.
     */
    public void receiveReward(){
        int wReward = Dice.weaponsReward();
        int sReward = Dice.shieldsReward();
        
        for(int i = 0; i < wReward; i++){
            Weapon wnew = newWeapon();
            receiveWeapon(wnew);
        }
        
        for(int i = 0; i < sReward; i++){
            Shield snew = newShield();
            receiveShield(snew);
        }
        
        int extraHealth = Dice.healthReward();
        setHealth(getHealth() + extraHealth);
        
    }
    
    @Override
    public String toString() {
        return "Nº: " + number + ", " + super.toString();
    }

    // Métodos privados

    /**
     * Recibe un arma y la agrega a la lista de armas del jugador.
     *
     * @param w Arma recibida.
     */
    private void receiveWeapon(Weapon w){
        for (int i=0; i<weapons.size(); i++){
            boolean discard = weapons.get(i).discard();
            if (discard)
                weapons.remove(weapons.get(i));
        }
        
        int size = weapons.size();
        if (size<MAX_WEAPONS)
            weapons.add(w);
    }

    /**
     * Recibe un escudo y lo agrega a la lista de escudos del jugador.
     *
     * @param s Escudo recibido.
     */
    private void receiveShield(Shield s){
        for (int i=0; i<shields.size(); i++){
            boolean discard = shields.get(i).discard();
            if (discard){
                shields.remove(shields.get(i));
            }
            
            int size = shields.size();
            if (size<MAX_SHIELDS){
                shields.add(s);
            }
        }
    }

    
    private Weapon newWeapon(){
        return weaponCardDeck.nextCard();
    }

    private Shield newShield(){
        return shieldCardDeck.nextCard();
    }
    

    /**
     * Calcula y devuelve la suma de las intensidades de las armas del jugador.
     *
     * @return La suma de las intensidades de las armas del jugador.
     */
    protected float sumWeapons(){
        float suma = 0.0f;
        for(Weapon weapon :  weapons){
            suma += weapon.attack();
        }
        return suma;
    }

    /**
     * Calcula y devuelve la suma de las intensidades de los escudos del jugador.
     *
     * @return La suma de las intensidades de los escudos del jugador.
     */
    protected float sumShields(){
        float suma = 0.0f;
        for(Shield shield : shields){
            suma += shield.protect();
        }
        return suma;
    }

    /**
     * Calcula y devuelve la energía defensiva total del jugador, que incluye la suma de los escudos y la inteligencia.
     *
     * @return La energía defensiva total del jugador.
     */
    protected float defensiveEnergy(){
        return sumShields() + getIntelligence();
    }

    /**
     * Gestiona el impacto recibido durante un combate, determinando si el jugador se defiende con éxito o sufre daño.
     *
     * @param receivedAttack La intensidad del ataque recibido.
     * @return true si el jugador se defiende con éxito, false si sufre daño.
     */
    private boolean manageHit(float receivedAttack){
        float defense = defensiveEnergy();
        
        if(defense < receivedAttack){
            gotWounded();
            incConsecutiveHits();
        }else{
            resetHits();
        }
        
        boolean lose;
        if((consecutiveHits == HITS2LOSE) || dead()){
            resetHits();
            lose = true;
        }else{
            lose=false;
        }
        return lose;
    }

    /**
     * Reinicia el contador de golpes consecutivos.
     */
    private void resetHits(){
        consecutiveHits = 0;
    }

    /**
     * Incrementa el contador de golpes consecutivos recibidos.
     */
    private void incConsecutiveHits(){
       consecutiveHits += 1; 
    }
}

File: IrrgartenRuby/irrgarten/player.rb
# encoding: utf-8
require_relative 'dice'
require_relative 'directions'
require_relative 'shield'
require_relative 'weapon'
require_relative 'labyrinth_character'

module Irrgarten
  class Player < LabyrinthCharacter
    @@MAX_WEAPONS = 2
    @@MAX_SHIELDS = 3
    @@INITIAL_HEALTH = 10
    @@HITS_TO_LOSE = 3

    attr_reader :number, :row, :col, :weapons, :shields, :consecutive_hits

    def initialize(number, intelligence, strength)
      super("Player #" + number.to_s, intelligence, strength, @@INITIAL_HEALTH)
      @number = number
      reset_hits
      @weapons = Array.new
      @shields = Array.new
    end

    #Falta el constructor de copia
    def copia(other)
      super(other)
      @weapons=other.weapons
      @shields=other.shields
      @consecutive_hits=other.consecutive_hits
      @number=other.number
    end
    
    def resurrect
    	@health = @@INITIAL_HEALTH
    	@@weapons = Array.new
      @shields = Array.new
      reset_hits
    end

    def get_number()
      @number
    end
    
    def move(direction, valid_moves)
      size = valid_moves.size
      contained = valid_moves.include?(direction)
      if (size > 0 && !contained)
          return valid_moves[0]
      else
          return direction
      end
    end


    def attack
      @strength + sum_weapons
    end

    def defend(received_attack)
      manage_hit(received_attack)
    end

    def receive_reward
      w_reward = Dice.weapons_reward
      s_reward = Dice.shields_reward

      w_reward.times do
        w_new = new_weapon
        receive_weapon(w_new)
      end

      s_reward.times do
        s_new = new_shield
        receive_shield(s_new)
      end

      extra_health = Dice.health_reward
      @health += extra_health
    end

    def to_s
      "Nº: #{@number}, Hits: #{@consecutive_hits}, " + super
    end

    private

    def receive_weapon(w)
      #lo borro si no es un arma y discard ==true
      @weapons.delete_if { |wi| wi.is_a?(Weapon) && wi.discard } 
      
      #Lo añado si hay espacio y es un arma
      if @weapons.size < @@MAX_WEAPONS && w.is_a?(Weapon)
        @weapons << w 
      end
    end

    def receive_shield(s)
      #lo borro si no es un escudo y discard ==true
      @shields.delete_if { |si| si.is_a?(Shield) && si.discard }
      
      #Lo añado si hay espacio y es un escudo
      if @shields.size < @@MAX_SHIELDS && s.is_a?(Shield)
        @shields << s 
      end
    end
    
    def new_weapon()
    	Weapon.new(Dice.weapon_power, Dice.uses_left)
    end
    
    def new_shield()
    	Shield.new(Dice.shield_power, Dice.uses_left)
    end

    def manage_hit(received_attack)
      defense = defensive_energy()
      if defense < received_attack
        got_wounded()
        inc_consecutive_hits()
      else
        reset_hits()
      end
      if (@consecutive_hits == @@HITS_TO_LOSE || dead())
        reset_hits()
        true
      else
        false
      end
    end

    def reset_hits
      @consecutive_hits = 0
    end

    def inc_consecutive_hits
      @consecutive_hits += 1
    end

    protected
    def sum_weapons
      fuerza = 0
      @weapons.each do |w|
        fuerza += w.attack
      end
      fuerza
    end

    def sum_shields
      proteccion = 0
      @shields.each do |s|
        proteccion += s.protect
      end
      proteccion
    end

    def defensive_energy
      @intelligence + sum_shields
    end

  end
end

